### PR TITLE
[Don't merge]Temporary fix for the issue

### LINKED ISSUE #41101: DDP should provide an option for not touching grad of locally unused params
## 🐛 Bug

For multi task training with different ranks running different tasks, parameters of inactive tasks are unused locally, even though used globally. By default DDP would allreduce these parameters and update all with the mean of grad across all ranks. Such grad update could trigger the weight decay with some optimizers (such as SGD) even for those inactive tasks on rank. Such weight update would propagate to ranks running active tasks and eventually pollute the grad there. 

## To Reproduce

1. Have two ranks R1 and R2. Both of them load the same model.
1. R1 runs task A branch and R2 runs task B branch.
1. After the first iteration, we would see B's grad on R1 and A's grad on R2 get updated.

## Expected behavior

B's grad on R1 and A's grad on R2 should never get updated as task B is inactive on R1 and task A is inactive on R2.

### PR DESCRIPTION
**Pull request by YinglinSun**

### CODE REVIEW

**Comment by ezyang** [comment]
remove the triaged tag if you need review
